File: src/main/java/core/procedures/generic/Get.java
package core.procedures.generic;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.Vector;
import core.utils.Utils;

import java.util.List;
import java.util.Map;
import java.util.Set;

public final class Get extends AFn {

  public Get() {
    super(new FnArgsBuilder().min(2).max(3).build());
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "get";
  }

  @Override
  public Object apply(Object... args) {
    Object defaultValue = null;
    if (args.length == 3) {
      defaultValue = args[2];
    }
    return get(args[0], args[1], defaultValue);
  }

  private Object get(Object col, Object key, Object defaultValue) {
    if (col instanceof Map) {
      return ((Map) col).getOrDefault(key, defaultValue);
    } else if ((col instanceof Map.Entry)) {
      if (Utils.INSTANCE.isInteger(key)) {
        int i = ((Number)key).intValue();
        if (i == 0) {
          return ((Map.Entry) col).getKey();
        } else if (i == 1) {
          return ((Map.Entry) col).getValue();
        }
      }
    } else if (col instanceof List) {
      if (Utils.INSTANCE.isInteger(key) && (((Number) key).intValue() < ((List) col).size())) {
        return ((List)col).get(((Number)key).intValue());
      }
    } else if (col instanceof Set) {
      if (((Set)col).contains(key)) {
        return key;
      }
    } else if (col instanceof Vector) {
      if (Utils.INSTANCE.isInteger(key) && (((Number) key).intValue() < ((Vector) col).size())) {
        return ((Vector)col).get(((Number)key).intValue());
      }
    } else if (col instanceof CharSequence) {
      if (Utils.INSTANCE.isInteger(key) && (((Number) key).intValue() < ((CharSequence) col).length())) {
        return ((CharSequence) col).charAt(((Number) key).intValue());
      }
    }
    return defaultValue;
  }
}

File: src/main/java/core/procedures/generic/Get.kt
package core.procedures.generic

import core.procedures.AFn
import core.scm.Vector
import core.utils.Utils

class Get : AFn<Any?, Any?>(name = "get", isPure = true, minArgs = 2, maxArgs = 3) {

    override operator fun invoke(vararg args: Any?) = invoke(args[0], args[1], args.getOrNull(2))

    override operator fun invoke(arg1: Any?, arg2: Any?, arg3: Any?): Any? {
        if (arg1 is Map<*, *>) {
            return (arg1 as Map<Any?, Any?>).getOrDefault(arg2, arg3)
        } else if (arg1 is Map.Entry<*, *>) {
            if (Utils.isInteger(arg2)) {
                val i = (arg2 as Number).toInt()
                when (i) {
                    0 -> return arg1.key
                    1 -> return arg1.value
                }
            }
        } else if (arg1 is List<*>) {
            if (Utils.isInteger(arg2) && (arg2 as Number).toInt() < arg1.size) {
                return arg1[arg2.toInt()]
            }
        } else if (arg1 is Set<*>) {
            if (arg1.contains(arg2)) {
                return arg2
            }
        } else if (arg1 is Vector) {
            if (Utils.isInteger(arg2) && (arg2 as Number).toInt() < arg1.size) {
                return arg1[arg2.toInt()]
            }
        } else if (arg1 is CharSequence) {
            if (Utils.isInteger(arg2) && (arg2 as Number).toInt() < arg1.length) {
                return arg1[arg2.toInt()]
            }
        }
        return arg3
    }
}
